File: packages/app/src/lib/binLookup.ts
import { detectNetwork } from '@cardledger/shared';
import type { Network } from '@cardledger/shared';

export interface BinInfo {
  network: Network | null;
  bank: string | null;
  variant: string | null;
}

function mapScheme(scheme?: string): Network | null {
  switch ((scheme ?? '').toLowerCase()) {
    case 'visa':
      return 'Visa';
    case 'mastercard':
      return 'Mastercard';
    case 'amex':
    case 'american express':
      return 'Amex';
    case 'rupay':
      return 'RuPay';
    default:
      return null;
  }
}

/**
 * Look up a BIN online (binlist.net), falling back to local network detection.
 * Never throws — always resolves to a BinInfo. Only the 6-digit BIN is sent.
 */
export async function lookupBin(bin: string): Promise<BinInfo> {
  const clean = bin.replace(/\D/g, '').slice(0, 6);
  const local: BinInfo = { network: detectNetwork(clean), bank: null, variant: null };
  if (clean.length < 6) return local;

  try {
    const res = await fetch(`https://lookup.binlist.net/${clean}`, {
      headers: { 'Accept-Version': '3' },
    });
    if (!res.ok) return local;
    const data = await res.json();
    const network = mapScheme(data?.scheme) ?? local.network;
    const bank: string | null = data?.bank?.name ?? null;
    const rawType = data?.type ? String(data.type) : '';
    const variant = rawType ? rawType.charAt(0).toUpperCase() + rawType.slice(1) : null;
    return { network, bank, variant };
  } catch {
    return local;
  }
}

File: packages/android-native/app/src/main/java/com/imvj/cardledger/data/net/BinLookup.kt
package com.imvj.cardledger.data.net

import android.util.Log
import com.imvj.cardledger.domain.detectNetwork
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.net.HttpURLConnection
import java.net.URL

private const val TAG = "BinLookup"
private const val CONNECT_TIMEOUT_MS = 4000
private const val READ_TIMEOUT_MS = 4000

data class BinInfo(val network: String?, val bank: String?, val variant: String?)

suspend fun lookupBin(bin: String): BinInfo = withContext(Dispatchers.IO) {
    val clean = bin.filter { it.isDigit() }.take(6)
    val local = BinInfo(detectNetwork(clean), null, null)
    if (clean.length < 6) return@withContext local
    try {
        val conn = (URL("https://lookup.binlist.net/$clean").openConnection() as HttpURLConnection).apply {
            setRequestProperty("Accept-Version", "3")
            connectTimeout = CONNECT_TIMEOUT_MS
            readTimeout = READ_TIMEOUT_MS
        }
        if (conn.responseCode != 200) return@withContext local
        val text = conn.inputStream.bufferedReader().use { it.readText() }
        val json = Json { ignoreUnknownKeys = true }.parseToJsonElement(text).jsonObject
        val scheme = json["scheme"]?.jsonPrimitive?.content
        val network = when (scheme?.lowercase()) {
            "visa" -> "Visa"; "mastercard" -> "Mastercard"
            "amex", "american express" -> "Amex"; "rupay" -> "RuPay"; else -> local.network
        }
        val bank = json["bank"]?.jsonObject?.get("name")?.jsonPrimitive?.content
        val type = json["type"]?.jsonPrimitive?.content?.replaceFirstChar { it.uppercase() }
        BinInfo(network, bank, type)
    } catch (e: Exception) {
        Log.w(TAG, "BIN lookup failed for $clean", e)
        local
    }
}
